llama-rs/src/lib.rs

```diff
@@ -1142,4 +1148,95 @@ impl Model {
 
         res
     }
+
+    /// Obtains a serializable snapshot of the current inference status. This
+    /// can be used to cache the state of the model and store them into a file.
+    ///
+    /// # Safety
+    ///
+    /// This function provides raw access to the underlying memory owned by the
+    /// ggml context. While the provided `LlamaModelMemory` object is alive, no
+    /// other methods for this model object should be called.
+    pub unsafe fn get_snapshot(&mut self) -> InferenceSnapshotRef<'_> {
+        use core::slice;
+        let memory_k = unsafe {
+            slice::from_raw_parts(self.memory_k.data() as *mut u8, self.memory_k.nbytes())
+        };
+        let memory_v = unsafe {
+            slice::from_raw_parts(self.memory_v.data() as *mut u8, self.memory_v.nbytes())
+        };
+
+        InferenceSnapshotRef {
+            npast: self.n_past,
+            memory_k,
+            memory_v,
+            last_n_tokens: self.last_n_tokens.clone(),
+            logits: self.last_logits.clone(),
+        }
+    }
+
+    /// Clears the model's state, so inference can be restarted without having
+    /// to load the weights again.
+    pub fn clear_state(&mut self) {
```

Comment:
Not a huge fan of this - it seems footgun-y (like it'd be easy to forget to clear the state after you're done). I'd suggest decoupling inference from the model itself, so that you have something like

```rust
let mut session: InferenceSession = model.infer();
session.feed_prompt("blah");
```
where all of the state relevant to this inference is in `InferenceSession`, and the model is purely data. (Bonus points if you have it borrow from `Model`.)

Reply:
Hmmmm, you're right :thinking: This requires moving the memory_k and memory_v tensors into another struct though. They should probably be owned by an entirely separate ggml context stored inside the `InferenceSession` to make things easier.